_posts/2023-07-12-React-unity-BDD.markdown

```diff
@@ -0,0 +1,560 @@
+
+```ts
+// featureFunctions.ts
+export async function selectAnimation(index: number) {
+  await waitFor(() => {
+    expect(screen.queryByTestId(`open-animation-${index}`)).not.toBeNull();
+  });
+
+  act(() => {
+    fireEvent.click(screen.queryByTestId(`open-animation-${index}`));
+  });
+}
+// Mock the Unity Response `returnedPlayerPositionAndPerspective`
+// So that callback is called and the scene is added
+
+async function getPlayerPositionUnityEvent(x: number, y: number, z: number) {
+  const event = JSON.stringify({
+    eventName: "receivedAnimationSceneDetails",
+    args: JSON.stringify({
+      playerPosition: { x: x, y: y, z: z },
+      screenshotData: "base64 image data",
+    }),
+  });
+  act(() => {
+    const handler = eventHandler(true, "");
+    handler(event);
+  });
+  await waitFor(() => {
+    expect(window.unityEvent).not.toBeUndefined();
+  });
+}
+
+export async function addNewScene(index?: number) {
+  await waitFor(() => {
+    expect(screen.queryByTestId("button-add-scene")).not.toBeNull();
+  });
+  act(() => {
+    fireEvent.click(screen.queryByTestId("button-add-scene")!);
+  });
+
+  await waitFor(() => {
+    // this would be our helper function to mock the unity response
+    // so now, each time we add a new scene the mocked player position is return and
+    // React app will simply update the app state with the returned data.
+    getPlayerPositionUnityEvent(50, 50, 120);
+  });
+}
+```
+
+#### Animation UI Component With Scene Cards
+
+Having defined the expected behavior for this feature, we can now proceed to implement the component that fulfills these requirements.
+
+```tsx
+// animationView.tsx
+import { CardActions, CardContent, styled, TextField, Typography } from "@mui/material";
+import React, { FC, useCallback } from "react";
+import AddButton from "../AddButton";
+// to understand the animation context provider please refer to the provided github repository
+import { setScenes, useAnimationDispatch, useAnimationState } from "../AnimationProvider";
+import { SceneCard } from "../SceneCard";
+
+const ViewAnimation: FC = () => {
+  const dispatch = useAnimationDispatch();
+  const { scenes } = useAnimationState();
+  const hasScenes = scenes.length > 0;
+
+  return (
+    <Stack data-testid="view-animation-details">
+      <AddButtonMenu />
+      <StyledCardContent sx={{ pt: 0, pb: hasScenes ? "4em" : undefined }}>
+        {updatedScenes.map(SceneWrapper)}
+      </StyledCardContent>
+    </Stack>
+  );
+};
+
+function SceneWrapper(scene: SceneDetails, index: number, scenes: SceneDetails[]) {
+  return (
+    <React.Fragment key={scene?.id || index}>
+      <SceneCard index={index} scene={scene} nextScene={scenes[index + 1]} />
+    </React.Fragment>
+  );
+}
+
+export default ViewAnimation;
```

Comment:
mmm... wayyyyyy too much elaborated code example this blog doesn't need, the example should be clear and concise.